### DISCUSSION TITLE
Javascript RPC Code Error: self signed certificate

### DISCUSSION THREAD

**Discussion by geokos009** [reach-sh/reach-lang#1335] [Q&A] [answered]
I am running a Javascript RPC code and it is showing this error:

2022-08-09T07:01:02.059Z DEBUG: I am alive
2022-08-09T07:01:02.904Z DEBUG: Attempting to process request by client ::ffff:172.23.0.1: POST /health {}
RPC /stdlib/parseCurrency [100]
internal/process/esm_loader.js:74
    internalBinding('errors').triggerUncaughtException(
                              ^
Error: self signed certificate
    at TLSSocket.onConnectSecure (_tls_wrap.js:1514:34)
    at TLSSocket.emit (events.js:375:28)
    at TLSSocket._finishInit (_tls_wrap.js:936:8)
    at TLSWrap.ssl.onhandshakedone (_tls_wrap.js:708:12) {
  code: 'DEPTH_ZERO_SELF_SIGNED_CERT'
}

The code seems correct and follows the example code in the reach github. What could be the reason for this error?

**Comment by jeapostrophe** [comment, marked as answer]
You have two options.

If you are using a standard RPC client, then you can disable TLS verification --- https://docs.reach.sh/rpc/#p_42

Or, you can give your RPC server a real TLS certificate --- https://docs.reach.sh/tool/#ref-usage-rpc-server

You should do the first if you are testing and the second if you are deploying for real.

**Reply by geokos009** [reply]
If I have to disable verification for testing then how do I do it using JS RPC code? There are no reach github examples showing how it is done.

For example this is how I do it in dart using the SDK:

final wallet = AlgorandWallet(provider: 'MyAlgo');

  final reach = Reach(
    host: '0.0.0.0',
    port: 3000,
    verify: false,
    timeout: 15,
    onSignTransaction: (transaction) async {
      return await wallet.signTransaction(transaction);
    },
  );

**Reply by jeapostrophe** [reply]
As this says --- https://docs.reach.sh/rpc/js/#js_mkRPC --- you pass the standard options --- https://docs.reach.sh/rpc/#p_42 --- when you construct it.

**Comment by geokos009** [comment]
I passed the following options as below:

const { rpc, rpcCallbacks } = await mkRPC({
  host: '0.0.0.0', 
  port: 3000, 
  verify: false, 
  timeout: 15, 
  key: 'opensesame',
});

I am getting the following error:

kend/node_modules/@reach-sh/rpc-client/index.mjs:20
      throw new Error(Neither \opts.${k}` nor ${defaults[k]} environment
            ^

Error: Neither opts.verify` nor undefined environment variable are configured!
    at file:///Users/georgekoshy/Documents/reach/oystro-with-oraclebackend/node_modules/@reach-sh/rpc-client/index.mjs:20:13
    at Array.forEach (<anonymous>)
    at mkRPC (file:///Users/georgekoshy/Documents/reach/oystro-with-oraclebackend/node_modules/@reach-sh/rpc-client/index.mjs:18:18)
    at file:///Users/georgekoshy/Documents/reach/oystro-with-oraclebackend/index.mjs:21:31
    at ModuleJob.run (internal/modules/esm/module_job.js:169:25)
    at async Loader.import (internal/modules/esm/loader.js:177:24)
    at async Object.loadESM (internal/process/esm_loader.js:68:5) 
    
What am I doing wrong?

**Reply by jeapostrophe** [reply]
You should set `verify` to `'0'`

**Reply by geokos009** [reply]
Thanks @jeapostrophe 

I tried it.

const { rpc, rpcCallbacks } = await mkRPC({
  host: '0.0.0.0', 
  port: 3000, 
  verify: '0', 
  timeout: 15, 
  key: 'opensesame',
});

const bal = await rpc('/stdlib/parseCurrency', 100);

I got the following error:

2022-08-22T04:34:49.937Z DEBUG: Attempting to process request by client ::ffff:172.19.0.1: POST /health {}
RPC /stdlib/parseCurrency [100]
internal/process/esm_loader.js:74
    internalBinding('errors').triggerUncaughtException(
                              ^

Error: self signed certificate
    at TLSSocket.onConnectSecure (_tls_wrap.js:1514:34)
    at TLSSocket.emit (events.js:375:28)
    at TLSSocket._finishInit (_tls_wrap.js:936:8)
    at TLSWrap.ssl.onhandshakedone (_tls_wrap.js:708:12) {
  code: 'DEPTH_ZERO_SELF_SIGNED_CERT'
}

**Reply by jeapostrophe** [reply]
This is very surprising to me, because the code is extremely simple --- https://github.com/reach-sh/reach-lang/blob/master/rpc-client/js/index.mjs#L29-L32

I don't understand why it doesn't work for you

**Comment by geokos009** [comment]
I noticed this message on 'import { mkRPC } from '@reach-sh/rpc-client'; when I hovered my cursor on the statement.

module "/Users/georgekoshy/Documents/reach/oystro-with-oraclebackend/node_modules/@reach-sh/rpc-client/index"
Could not find a declaration file for module '@reach-sh/rpc-client'. '/Users/georgekoshy/Documents/reach/oystro-with-oraclebackend/node_modules/@reach-sh/rpc-client/index.mjs' implicitly has an 'any' type.
  Try `npm i --save-dev @types/reach-sh__rpc-client` if it exists or add a new declaration (.d.ts) file containing `declare module '@reach-sh/rpc-client';`ts(7016)

What should I do? Perhaps this may solve the issue

**Reply by jeapostrophe** [reply]
I think this error is saying "I can't find the types of this", but I don't really know. I think you might have TypeScript configured to error when there's an `any` type and you can get rid of the error by not doing that... but I don't really know.

**Reply by geokos009** [reply]
I found these 2 links : 

https://stackoverflow.com/questions/56927551/how-to-resolve-could-not-find-a-declaration-file-for-module-request-context

https://stackoverflow.com/questions/41292559/could-not-find-a-declaration-file-for-module-module-name-path-to-module-nam/54417184#54417184

What direction should I take?

**Reply by jeapostrophe** [reply]
Can you look at this @DanBurton ?

**Reply by DanBurton** [reply]
The error's suggestion is:

> Try npm i --save-dev @types/reach-sh__rpc-client if it exists or add a new declaration (.d.ts) file containing declare module '@reach-sh/rpc-client';ts(7016)

Note that `@types/reach-sh__rpc-client` does not exist so you should take the latter suggestion of making a new declaration file. Both links that you found seem to be doing that.
